smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
-        else:
-            oob = (numpy.any(x < self.domain[0]) or 
-                    numpy.any(x > self.domain[1]))
+        # validate inputs
+        X = numpy.array(X, ndmin=2)
+        if X.shape[1] != self.num_dimensions:
+            raise IndexError("Must be 2D array with shape (n_samples, n_features)")
+        if not self._valid_cache:
+            raise RuntimeError("Model must be fit!")
+        oob = any(
+            numpy.any(X[:, i] < self.domain[i][0])
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
         if oob:
-            raise ValueError('x must lie in domain of surrogate')
-        # transform point into basis domain and evaluate
-        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
-        value = 0
-        for coeff, basis in zip(self._coefficients, self.grid.basis_functions):
-            if self.dimension > 1:
-                term = numpy.prod(
-                        [basis[i](x_scaled[...,i]) for i in range(len(basis))],
-                        axis=0)
-            else:
-                term = basis(x_scaled)
-            value += coeff*term
+            raise ValueError("X must lie in domain of surrogate")
+
+        # create lookup table
+        num_basis_max = numpy.max([len(p) for p in self._basis_sets])
+        if any(
+            any(bf._is_complex for bf in basis_set) for basis_set in self.basis_sets
+        ):
+            lookup_table = numpy.zeros(
+                (self.num_dimensions, num_basis_max, len(X)), dtype="complex_"
+            )
+        else:
+            lookup_table = numpy.zeros(
+                (self.num_dimensions, num_basis_max, len(X)), dtype="complex_"
+            )
+        # solve for the inputs at all the basis functions
+        for dim in range(self.num_dimensions):
+            for i, basis_fun in enumerate(self.basis_sets[dim]):
+                new_X = basis_fun.domain[0] + (
+                    basis_fun.domain[1] - basis_fun.domain[0]
+                ) * (
+                    (X[:, dim] - self.domain[dim, 0])
+                    / (self.domain[dim, 1] - self.domain[dim, 0])
+                )
+                numpy.clip(new_X, basis_fun.domain[0], basis_fun.domain[1], out=new_X)
```

Comment:
Two things here:
1. I think we should enforce that all functions in a `BasisFunctionSet` must have the same domain. That way, you can scale the coordinates once per dimension, rather than once per basis function, which will speed up evaluation alot. You can do something like assert all domains are the same, then add a property that returns the domain of the first function in the set.
2. I think we should add a helper method to the `BasisFunction` (with a pass through method on the `BasisFunctionSet` too) that scales points from a domain to the domain of the basis function (set).

Reply:
> I think we should add a helper method to the `BasisFunction` (with a pass through method on the `BasisFunctionSet` too) that scales points from a domain to the domain of the basis function (set).
I see why it makes sense to add a helper method to the `BasisFunctionSet` for that, but not the `BasisFunction`. We can just have `BasisFunctionSet` know how to scale the points and get the right domain from the first element in `basis_functions`.

Reply:
Though if either the BasisFunction or the BasisFunctionSet handle the scaling, they should also handle the evaluations and the derivative since the derivative has to be multiplied by a constant related to scaling the domain.

Reply:
> I see why it makes sense to add a helper method to the `BasisFunctionSet` for that, but not the `BasisFunction`. We can just have `BasisFunctionSet` know how to scale the points and get the right domain from the first element in `basis_functions`.

My reasoning was that it is the same amount of work to put this method on one of the basis functions and the basis function is what actually owns the domain, and there could be use cases where someone wanted to map the inputs for a single basis function. If the domains are the same for all functions in the set, the set can call the transform method of any basis function (but not vice versa).

> Though if either the BasisFunction or the BasisFunctionSet handle the scaling, they should also handle the evaluations and the derivative since the derivative has to be multiplied by a constant related to scaling the domain.

I don't think this necessarily has to be the case. If the transform is clearly defined, the caller can be responsible for applying the scale factor. However, I can see it being convenient to evaluate all of the functions in a set (or their derivatives) as this is our most common use case, so I am on board if you want to implement these and use them in the code. It would probably help reduce some duplication.

Reply:
`_scale_to_domain` is now implemented in `BasisFunction` and `BasisFunctionSet`, and the `BasisFunctionSet` now has methods for evaluating all the functions in the set and all their derivatives. The new methods in the `BasisFunctionSet` have reduced code duplication significantly. In the surrogate methods generating the lookup tables now takes one for loop.

Reply:
The basis function now has a _scale_to_domain function